src/common/proxy_wasm/ngx_proxy_wasm_foreign_call.c

```diff
@@ -0,0 +1,293 @@
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+#if (NGX_WASM_LUA)
+#include <ngx_wasm_lua_resolver.h>
+#endif
+#include <ngx_proxy_wasm_foreign_call.h>
+
+
+void
+ngx_proxy_wasm_foreign_call_destroy(ngx_proxy_wasm_foreign_call_t *call)
+{
+    ngx_pfree(call->pwexec->pool, call);
+}
+
+
+#if (NGX_WASM_HTTP)
+#if (NGX_WASM_LUA)
+static void
+ngx_proxy_wasm_foreign_call(ngx_proxy_wasm_foreign_call_t *call)
+{
+    ngx_proxy_wasm_exec_t  *pwexec = call->pwexec;
+    ngx_proxy_wasm_err_e    ecode = NGX_PROXY_WASM_ERR_NONE;
+    ngx_proxy_wasm_step_e   step = pwexec->parent->step;
+
+    ngx_queue_remove(&call->q);
+
+    pwexec->foreign_call = call;
+
+    pwexec->parent->phase = ngx_wasm_phase_lookup(&ngx_http_wasm_subsystem,
+                                                  NGX_WASM_BACKGROUND_PHASE);
+
+    ecode = ngx_proxy_wasm_run_step(pwexec,
+                                    NGX_PROXY_WASM_STEP_FOREIGN_CALLBACK);
+    if (ecode != NGX_PROXY_WASM_ERR_NONE) {
+        /* TODO: error handling */
+    }
+
+    pwexec->parent->step = step;
+    pwexec->foreign_call = NULL;
+
+    if (ngx_proxy_wasm_foreign_calls_total(pwexec)) {
+        ngx_log_debug0(NGX_LOG_DEBUG_WASM, pwexec->log, 0,
+                       "proxy_wasm more foreign function callbacks pending...");
+
+        ngx_wasm_yield(&call->rctx->env);
+        ngx_proxy_wasm_ctx_set_next_action(pwexec->parent,
+                                           NGX_PROXY_WASM_ACTION_PAUSE);
+
+    } else {
+        ngx_log_debug0(NGX_LOG_DEBUG_WASM, pwexec->log, 0,
+                       "proxy_wasm last foreign function callback handled");
+
+        ngx_wasm_continue(&call->rctx->env);
+        ngx_proxy_wasm_ctx_set_next_action(pwexec->parent,
+                                           NGX_PROXY_WASM_ACTION_CONTINUE);
+
+        /* resume current step if unfinished */
+        ngx_proxy_wasm_resume(pwexec->parent, pwexec->parent->phase, step);
+    }
+
+    ngx_proxy_wasm_foreign_call_destroy(call);
+}
+
+
+static ngx_int_t
+ngx_proxy_wasm_foreign_call_buffer_create(ngx_proxy_wasm_foreign_call_t *call,
+    size_t size)
+{
+    ngx_chain_t              *cl;
+    ngx_proxy_wasm_exec_t    *pwexec = call->pwexec;
+    ngx_http_wasm_req_ctx_t  *rctx = call->rctx;
+
+    ngx_wa_assert(pwexec);
+    ngx_wa_assert(size);
+
+    /* TODO: if size > NGX_MAX_ERROR_STR, split allocation into N buffers */
+
+    ngx_wa_assert(size <= NGX_MAX_ERROR_STR);
+
+    cl = ngx_wasm_chain_get_free_buf(pwexec->pool,
+                                     &rctx->free_bufs,
+                                     size, buf_tag,
+                                     rctx->sock_buffer_reuse);
```

Comment:
Great to see this all the updates; also to see such a shorter function. In this case though I think we have to hard-code the reuse argument to `1`, as `rctx->sock_buffer_reuse` is meant to be tied to TCP/UDP sockets. It is currently only used for HTTP dispatch sockets while all other calls to `ngx_wasm_chain_get_free_buf` are hard-coded as `1`. We could change this later or tie the hard-coded values to another directive (I doubt it's worth it, but we could), for now let's keep it that way and hard-code the reuse argument here?

Reply:
I can also make the change myself, but just an fyi

Reply:
Thanks for taking the time to explain that. Will had-code it to `1`.